src/_pytest/nodes.py

```diff
@@ -484,38 +497,47 @@ def _check_initialpaths_for_relpath(session, fspath):
 
 
 class FSCollector(Collector):
-    def __init__(
-        self,
-        fspath: py.path.local,
-        parent=None,
-        config: Optional[Config] = None,
-        session: Optional["Session"] = None,
+    @classmethod
+    def from_parent(
+        cls,
+        parent: Node,
+        *,
+        fspath: Optional[py.path.local] = None,
+        fs_path: Optional[Path] = None,
         nodeid: Optional[str] = None,
-    ) -> None:
-        name = fspath.basename
-        if parent is not None:
+        name: Optional[str] = None,
```

Comment:
`name` being `""` is not valid right? If so I think it would be better to have this as: `name: str = ""` to signal that a name was not given; as it stands now it is possible to pass `name=""` to `from_parent` and end up with an empty name.

Reply:
name as `""` is done by session :unamused: 

we might be able to do it like that tho

Reply:
Ahh right.

If it isn't too much work I think it is worth it, as it will make for a better external API.

Reply:
i reiterated the idea, there is no sane/safe concept of "no Name" atm, 
name not given should be as None, to use a distinct type, empty name should be a error